### PR TITLE
Messaging system implementation

### PR DESCRIPTION
**Pull request by cmyui**
Still needs testing before merge

### CODE REVIEW

File: rgdps/api/gd/messages.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+from rgdps.common.validators import Base64String
+from rgdps.constants.errors import ServiceError
+from rgdps.models.message import MessageDirection
+from rgdps.models.user import User
+from rgdps.usecases import messages
+
+
+async def message_post(
+    ctx: HTTPContext = Depends(),
+    user: User = Depends(authenticate_dependency()),
```
Comment (cmyui) [inline rgdps/api/gd/messages.py]: privs might be wrong

File: rgdps/api/gd/messages.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        subject=subject,
+        content=content,
+    )
+
+    return responses.success()
+
+
+async def messages_get(
+    ctx: HTTPContext = Depends(),
+    user: User = Depends(authenticate_dependency()),
```
Comment (cmyui) [inline rgdps/api/gd/messages.py:51]: privs might be wrong

File: rgdps/usecases/messages.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+async def create(
+    ctx: Context,
+    sender_user_id: int,
+    recipient_user_id: int,
+    subject: str,
+    content: str,
+) -> Message | ServiceError:
+    if not await repositories.user.from_id(ctx, recipient_user_id):
+        return ServiceError.MESSAGES_INVALID_RECIPIENT
```
Comment (cmyui) [inline rgdps/usecases/messages.py:43]:
```suggestion
```

File: rgdps/usecases/messages.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
```
Comment (cmyui) [inline rgdps/usecases/messages.py:44]:
```suggestion
```

Comment (cmyui) [inline rgdps/usecases/messages.py]:
```suggestion
    recipient = await repositories.user.from_id(ctx, recipient_user_id):
    if recipient is None:
        return ServiceError.MESSAGES_INVALID_RECIPIENT
```

File: rgdps/usecases/messages.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    recipient_user_id: int,
+    subject: str,
+    content: str,
+) -> Message | ServiceError:
+    recipient = await repositories.user.from_id(ctx, recipient_user_id):
+    if recipient is None:
+        return ServiceError.MESSAGES_INVALID_RECIPIENT
+    recipient = await repositories.user.from_id(ctx, user_id=recipient_user_id)
+    if recipient is None:
+        return ServiceError.USER_NOT_FOUND
```
Comment (cmyui) [inline rgdps/usecases/messages.py]:
```suggestion
```

File: database/migrations/011_messages_table.up.sql
Comment (RealistikDash) [inline database/migrations/011_messages_table.up.sql:1]: The migration here will require a rename to the next available free number.

**Comment by RealistikDash** [comment]
Btw, all of the pre-requesites for this have been completed a while ago. Do you wish to finish/test this PR or should I take it over?

File: rgdps/api/gd/messages.py (hunk 1 above)
Comment (RealistikDash) [reply rgdps/api/gd/messages.py]: Should use `MESSAGES_SEND`

File: rgdps/api/gd/messages.py (hunk 2 above)
Comment (RealistikDash) [reply rgdps/api/gd/messages.py:51]: No privileges required here (yet)

**Comment by cmyui** [comment]
> Btw, all of the pre-requesites for this have been completed a while ago. Do you wish to finish/test this PR or should I take it over?

Hey, sorry I have very little time atm - end of quarter crunch at work is pretty wild - so I think you might need to take over :pray:

File: database/migrations/011_messages_table.up.sql
Comment (lenforiee) [reply database/migrations/011_messages_table.up.sql:1]: resolved in d88272d

File: rgdps/usecases/messages.py [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+
+    if message is None:
+        return ServiceError.MESSAGES_NOT_FOUND
+
+    return message
+
+
+async def delete(
+    ctx: Context,
+    user: User,
```
Comment (RealistikDash) [inline rgdps/usecases/messages.py]: I would probably just pass user_id here.

File: rgdps/usecases/messages.py [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+
+    if message is None:
+        return ServiceError.MESSAGES_NOT_FOUND
+
+    return message
+
+
+async def delete_by_user(
+    ctx: Context,
+    user: User,
```
Comment (RealistikDash) [inline rgdps/usecases/messages.py]: I would probably just pass user_id here.

File: rgdps/usecases/messages.py [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+        subject=subject,
+        content=content,
+    )
+
+    return message
+
+
+async def mark_message_as_seen(
+    ctx: Context,
+    user: User,
```
Comment (RealistikDash) [inline rgdps/usecases/messages.py]: I would probably just pass user_id here.

File: rgdps/usecases/messages.py [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+
+
+class PaginatedMessagesResponse(NamedTuple):
+    messages: list[MessageResponse]
+    total: int
+
+
+async def from_id(
+    ctx: Context,
+    user: User,
```
Comment (RealistikDash) [inline rgdps/usecases/messages.py]: I would probably just pass user_id here.

File: rgdps/usecases/messages.py [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+    message: Message
+    user: User
+
+
+class PaginatedMessagesResponse(NamedTuple):
+    messages: list[MessageResponse]
+    total: int
+
+
+async def from_id(
```
Comment (RealistikDash) [inline rgdps/usecases/messages.py]: I would probably call this something along the lines of `get` for consistency.

File: rgdps/usecases/messages.py [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+    else:
+        recipient = await repositories.user.from_id(ctx, message.sender_user_id)
+
+    if recipient is None:
+        return ServiceError.MESSAGES_INVALID_OWNER
+
+    return MessageResponse(message=message, user=recipient)
+
+
+async def from_recipient_user_id(
```
Comment (RealistikDash) [inline rgdps/usecases/messages.py]: I would robably name this something along the lines of `get_sent`

**Review by RealistikDash — CHANGES_REQUESTED** [review]

File: rgdps/usecases/messages.py (hunk 9 above)
Comment (lenforiee) [reply rgdps/usecases/messages.py]: resolved in 13c7f80

File: rgdps/usecases/messages.py (hunk 8 above)
Comment (lenforiee) [reply rgdps/usecases/messages.py]: resolved in 13c7f80

File: rgdps/usecases/messages.py (hunk 7 above)
Comment (lenforiee) [reply rgdps/usecases/messages.py]: resolved in 13c7f80

File: rgdps/usecases/messages.py (hunk 6 above)
Comment (lenforiee) [reply rgdps/usecases/messages.py]: resolved in 13c7f80

File: rgdps/usecases/messages.py (hunk 11 above)
Comment (lenforiee) [reply rgdps/usecases/messages.py]: resolved in 8109c21

File: rgdps/usecases/messages.py (hunk 10 above)
Comment (lenforiee) [reply rgdps/usecases/messages.py]: resolved in 8109c21

**Comment by RealistikDash** [comment]
Tested, works well.

**Review by RealistikDash — APPROVED** [review]
Looks good

**Pull request merged**